mw_basic_auth.go

```diff
@@ -117,6 +111,66 @@ func (k *BasicAuthKeyIsValid) ProcessRequest(w http.ResponseWriter, r *http.Requ
 		ctxSetAuthToken(r, keyName)
 	}
 
-	// Request is valid, carry on
 	return nil, http.StatusOK
 }
+
+func (k *BasicAuthKeyIsValid) handleAuthFail(w http.ResponseWriter, r *http.Request, token string) (error, int) {
+
+	// Fire Authfailed Event
+	AuthFailed(k, r, token)
+
+	// Report in health check
+	reportHealthValue(k.Spec, KeyFailure, "-1")
+
+	return k.requestForBasicAuth(w, "User not authorised")
+}
+
+func (k *BasicAuthKeyIsValid) compareHashAndPassword(hash string, password string, logEntry *logrus.Entry) error {
+
+	cacheEnabled := !k.Spec.BasicAuth.DisableCaching
+	cacheDuration := time.Duration(60) * time.Second
+	if k.Spec.BasicAuth.CacheTTL > 0 {
+		cacheDuration = time.Duration(k.Spec.BasicAuth.CacheTTL) * time.Second
+	}
+
+	hasher := murmur3.New32()
+	passwordBytes := []byte(password)
+
+	doBcrypt := func(cacheEnabled bool, cacheDuration time.Duration, hashedPassword string, password []byte) error {
+		if err := bcrypt.CompareHashAndPassword([]byte(hash), password); err != nil {
+
+			logEntry.Debug("bcrypt: failed auth")
+			return err
+		}
+
+		if cacheEnabled {
+
+			logEntry.Debug("cache enabled: save")
+			k.cache.Set(hashedPassword, string(hasher.Sum([]byte(password))), cacheDuration)
+		}
+
+		return nil
+	}
+
+	if !cacheEnabled {
+
+		logEntry.Debug("cache disabled")
+		return doBcrypt(cacheEnabled, cacheDuration, hash, passwordBytes)
+	}
+
+	cachedPass, inCache := k.cache.Get(hash)
+	if !inCache {
+
+		logEntry.Debug("cache enabled: miss: bcrypt failover")
+		return doBcrypt(cacheEnabled, cacheDuration, hash, passwordBytes)
+	}
+
+	if cachedPass.(string) != string(hasher.Sum([]byte(password))) {
```

Comment:
I think this needs to be `cachedPass.(string) != string(hasher.Sum(nil))`

`hasher.Sum` doesn't calculate hash, it returns current state of hash (which needs to be calculated with method `hasher.Sum` earlier) with passed bytes appended to result